single_number/single_number.py

```diff
@@ -1,12 +1,38 @@
 '''
 Input: a List of integers where every int except one shows up twice
 Returns: an integer
+
+Non-empty array
+Each number appears twice except one, which appears once
+Return the number that appears once
+Based on examples, array is not sorted
+Algorithm should do a single pass through the array
+Space complexity O(1)
+
 '''
 def single_number(arr):
-    # Your code here
+    # Itereate through and call the "not in" method for each element
+    # enumerate the list and call "not in" on the list before that element
+    # and after that element. If both are true, that's the number we return
 
-    pass
+    for index, element in enumerate(arr):
+        # special case for first element
+            # don't look backward, just forward
+        if index == 0:
+            if element not in arr[index + 1:]:
+                return element
+        # otherwise, check the first part of the list, then the second part
+        elif element not in arr[:index]:
+            if element not in arr[index + 1:]:
+                return element
+    
 
+    '''
+    My question with this solution, is if this is going through
+    the array only once.  The complexity is probably something like
+    O(nlogn) over the array, depending on Python's "not in" implementation.
+    Nothing was stated in the instructions about the time complexity.
+    '''
```

Comment:
So, `not in` or any type of check for an element in an iterable object is `O(n)` because you need to check every element in the iterable.

In both your `if` and your `elif`, you're using `not in` to check every element before and after `index`. So in both of those cases the run-time is `O(n - 1)`.

If you factor in the outer for-loop `O(n)`, your total runtime is `O(n * n - 1)` or `O(n^2)`.

There's a way to achieve an `O(n)` run-time complexity, but an `O(n)` space complexity. The hint here is to use dictionaries.

But, your solution doesn't use any auxiliary space, so your space complexity is `O(1)` which was the objective of the challenge, so consider this a victory!

That being said, there is a way to achieve `O(n)` run-time and `O(1)` space complexity, but the solution is not obvious; not in the slightest bit. If you're interested in trying to find the solution, check out bitwise operators in python.